### PR TITLE
Fix padding side

### PR DESCRIPTION
**Pull request by thomasw21**
Fixes: #29 

As @kkawamu1 noticed, `<pad>` between `inputs` and `targets` is bad. Right now I've chose to use `padding_side="left"` as a fix. The current fix does:
```
<INPUT> <INPUT_PAD> <TARGET> <TARGET_PAD> -> <INPUT_PAD> <INPUT> <TARGET> <TARGET_PAD>
```

 The more permanent fix is to to do:
```
<INPUT> <INPUT_PAD> <TARGET> <TARGET_PAD> -> <INPUT> <TARGET> <INPUT_PAD> <TARGET_PAD>
```
This becomes especially useful when you start to leave the `text input | text target` paradigm so multiple step one.

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by VictorSanh — COMMENTED** [review]
that mostly looks good to me! thank you @thomasw21 !
left a few questions before merging.

I imagine you were able to replicate numbers both for encoderdecoder and decoder only?

File: t0/model.py [hunk 1]
```diff
@@ -81,16 +81,18 @@ def __init__(self, config, model_name_or_path: Optional[str], **kwargs):
...
             "attention_mask": torch.cat([batch["attention_mask"], batch["labels_attention_mask"]], dim=-1),
         }
         # Set position ids correctly to take care of padding tokens between inputs_ids and labels
-        # Empty attention_mask is a forbidden value, ie full of zeros. In fact the first element should be 1 as the input
-        #   cannot be empty
-        assert torch.all(model_inputs["attention_mask"][:,0] == 1), "First element in the attention mask should be 1."
-        position_ids = torch.cumsum(model_inputs["attention_mask"].to(torch.long), dim=-1) - 1
+        position_ids = torch.maximum(
+            torch.cumsum(model_inputs["attention_mask"].to(torch.long), dim=-1) - 1,
+            torch.zeros(1, dtype=torch.long, device=device)[None, None]
```
[the remark this answers is not in the document]
Comment (thomasw21) [reply t0/model.py:94]: Depends on how your model is robust to receiving negative positional embeddings. If the first tokens is a masked value, then the first position embedding is `-1` and so passing that value to the model might crash. The other work around would be something like `position_ids = (torch.cumsum() * model_inputs["attention_mask"]).to(torch.long)`

[the remark this answers is not in the document]
Comment (VictorSanh) [reply t0/model.py:94]: yes i am stupid, makes sense! thanks!

File: evaluation/run_eval.py [hunk 2]
```diff
@@ -265,8 +265,9 @@ def preprocess_function(examples):
         tokenized_targets = [
             tokenizer(
                 ans_choi,
-                padding=True,
-                max_length=args.target_max_length,
+                # padding is on the right here.
+                padding=False,
```
[the remark this answers is not in the document]
Comment (thomasw21) [reply evaluation/run_eval.py:269]: Hum I don't think I understand. Are all answer choices the same number of tokens? Seems very dependent on the tokenizer and so I'd expect that in some cases there is actual padding.

[the remark this answers is not in the document]
Comment (VictorSanh) [reply evaluation/run_eval.py:269]: > Are all answer choices the same number of tokens?

No. You can have `acceptable` and `not acceptable` as answer choices. So that's why I am confused to see `padding=False`. I would think we want to pad on the right.

[the remark this answers is not in the document]
Comment (thomasw21) [reply evaluation/run_eval.py:269]: Ah okay I understand better the question. Padding is done manually. https://github.com/bigscience-workshop/t-zero/blob/cf4a6781e4b8a0c5c5e45e40363a3beee0282318/t0/data_collator.py#L83-L90

**Comment by VictorSanh** [comment]
good to go! feel free to merge @thomasw21 !
